### DISCUSSION TITLE
clear email_activate after $user->activate();

### DISCUSSION THREAD

**Discussion by mshannaq** [codeigniter4/shield#966] [Q&A]
is there any way to clear `email_activate` for the user which is on database table `auth_identities`

![image](https://github.com/codeigniter4/shield/assets/162535/30eabbf8-da03-4d27-a81a-f6c1d305cbd6)

because even if I activate the user email using:

```php

            if (! $user->isActivated()) {
                $user->activate();
                $response['status'] = 'activated';
            } else {
                $response['error'] = lang("Users.UserIsAlreadyActivated");
            }
           return $response;

```
and the user try to login it will be asked to activate email account while his email is already activated if there is `email_activate` row in database table.

![image](https://github.com/codeigniter4/shield/assets/162535/c528efd9-9f51-4412-94ea-08b231a441c5)

while I make sure fro the databse and the user looks it is active:
![image](https://github.com/codeigniter4/shield/assets/162535/1db643bb-494f-47cd-a78b-630c953be8d8)

but because he has row in auth_identities for `email_activate` it still ask for token

how to remove this raw after activate the user account when running the code above.

is there any Available method for that in the entity or what?

Thanks.

**Comment by mshannaq** [comment]
Hello

I used `UserIdentityModel` by modify my controller code from

```php
if (! $user->isActivated()) {
    $user->activate();
    $response['status'] = 'activated';
} else {
    $response['error'] = lang("Users.UserIsAlreadyActivated");
}
return $response;
```

into:

```php
if (! $user->isActivated()) {
    $user->activate();
    $UserIdentityModel = new UserIdentityModel();
    $UserIdentityModel->deleteIdentitiesByType($user,"email_activate");
    $response['status'] = 'activated';
} else {
    $response['error'] = lang('Users.UserIsAlreadyActivated');
}
return $response;
```

I use `$UserIdentityModel->deleteIdentitiesByType($user,"email_activate");` to delete `email_activate` type for the user when activating email. and it works fine

is that best the way to do that? or there is another best way to do it?

Thanks

**Reply by kenjis** [reply]
I don't know what you want to do.
Why you create user identity for activation, and after it delete it?

**Reply by mshannaq** [reply]
@kenjis I did not create user identity 

I Just create a new user using
`php spark shield:user create`
and before activate the user using `php spark shield:user activate` or extenally by my code.

I try to login with that user and shield automatically create `email_activate` identity.

and If I try to activate the user account. the `email_activate` still in the database and will not allow user to login until he enter the code that sent to email even If his email status `active` = `1`.

so I was query about how to remove `email_activate` identity unti I did it using:

```php
if (! $user->isActivated()) {
    $user->activate();
    $UserIdentityModel = new UserIdentityModel();
    $UserIdentityModel->deleteIdentitiesByType($user,"email_activate");
    $response['status'] = 'activated';
} else {
    $response['error'] = lang('Users.UserIsAlreadyActivated');
}
return $response;
```

**Reply by kenjis** [reply]
I still don't understand your situation.

Why do you need to login using a user without activation?
If you don't need email activation, why don't you run `php spark shield:user activate` soon?

Or if you don't need email activation, why did you configure to require it?

**Reply by mshannaq** [reply]
it was a complicated situation ..

how to disable require for activation at all? is there any way to do that?

**Reply by kenjis** [reply]
By default, Shield does not require email activation.
https://codeigniter4.github.io/shield/references/authentication/auth_actions/
